src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp

```diff
@@ -1285,28 +1503,108 @@ void DiscoveryDataBase::add_writer_to_topic_(
         const eprosima::fastrtps::rtps::GUID_t& writer_guid,
         const std::string& topic_name)
 {
-    // Update writers by topic
-    std::map<std::string, std::vector<eprosima::fastrtps::rtps::GUID_t>>::iterator topic_it =
+    /* Manage adding a virtual writer */
+    bool virtual_topic_is_new = true;
+    // If the topic is virtual, add  the writer to all existing topics
+    if (topic_name == virtual_topic_)
+    {
+        // Add writer to all topics
+        for (auto topic: writers_by_topic_)
+        {
+            // Check if the writer is already there before adding it
+            auto topic_it = std::find(topic.second.begin(), topic.second.end(), writer_guid);
+            if (topic_it == topic.second.end())
+            {
+                logInfo(DISCOVERY_DATABASE,
+                    "Adding virtual writer " << writer_guid << " to writers_by_topic_[ " << topic_name << "]");
+                topic.second.push_back(writer_guid);
+            }
+            // We should never try to add a new entry if the writer was already in the topic
+            else
+            {
+                logWarning(DISCOVERY_DATABASE,
+                    "Trying to add already present virtual writer " << writer_guid << " to writers_by_topic_[ "
+                    << topic_name << "]. Ignored");
+            }
+
+            // The virtual topic already existed, so we don't need to create the entry for it later on
+            if (topic.first == virtual_topic_)
+            {
+                virtual_topic_is_new = false;
+            }
+        }
+
+        // There is no virtual_topic_ entry in writers_by_topic_. Therefore, we need to create one and add this virutal
+        // writer
+        if (virtual_topic_is_new)
+        {
+            // Try to insert the writer
+            auto retu = writers_by_topic_.insert(
+                std::pair<std::string, std::vector<fastrtps::rtps::GUID_t>>(
+                    topic_name,
+                    std::vector<fastrtps::rtps::GUID_t>({writer_guid})
+                )
+            );
+            if (!retu.second)
+            {
+                logError(DISCOVERY_DATABASE, "Could not insert writer " << writer_guid << " in topic " << topic_name);
+            }
+            else
+            {
+                logInfo(DISCOVERY_DATABASE, "New topic in writers_by_topic_: " << topic_name);
+            }
+        }
+        // If the topic was virtual, then there is no extra work to do
+        return;
+    }
+
+    /* Update writers by topic */
+    // At this point we know that the topic is real (NOT virutal)
+    // Find the topic_name entry in writers_by_topic_
+    std::map<std::string, std::vector<eprosima::fastrtps::rtps::GUID_t> >::iterator topic_it =
             writers_by_topic_.find(topic_name);
-    if (topic_it != writers_by_topic_.end())
+
+    // There was not topic_name entry in writers_by_topic_, so the topic is new
```

Comment:
This is gonna be tricky. But whenever we create a topic in writer or reader, we must create the same topic in the opposite one, so the virtual endpoints can be in it as well.
This could means to create extra functions that encapsulates some functionality of these ones so they can be called indifferently with a normal or a virtual topic (depends on the design).